nexus/src/app/instance.rs

```diff
@@ -2290,4 +2366,91 @@ mod tests {
+            start: false,
+        };
+
+        let instance_id = InstanceUuid::from_untyped_uuid(Uuid::new_v4());
+        let project_id = Uuid::new_v4();
+        let instance = DbInstance::new(instance_id, project_id, &params);
+
+        let propolis_id = PropolisUuid::from_untyped_uuid(Uuid::new_v4());
+        let sled_id = SledUuid::from_untyped_uuid(Uuid::new_v4());
+        let vmm = DbVmm::new(
+            propolis_id,
+            instance_id,
+            sled_id,
+            ipnetwork::IpNetwork::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 0)
+                .unwrap(),
+            0,
+            VmmInitialState::Starting,
+        );
+
+        (instance, vmm)
+    }
+
+    #[test]
+    fn test_instance_start_allowed_when_no_vmm() {
+        let logctx = test_setup_log("test_instance_start_allowed_when_no_vmm");
+        let (mut instance, _vmm) = make_instance_and_vmm();
+        instance.runtime_state.nexus_state = DbInstanceState::NoVmm;
+        assert!(instance_start_allowed(&logctx.log, &instance, &None).is_ok());
+    }
+
+    #[test]
+    fn test_instance_start_allowed_when_vmm_in_saga_unwound() {
+        let logctx = test_setup_log(
+            "test_instance_start_allowed_when_vmm_in_saga_unwound",
+        );
+        let (mut instance, mut vmm) = make_instance_and_vmm();
+        instance.runtime_state.nexus_state = DbInstanceState::Vmm;
+        instance.runtime_state.propolis_id = Some(vmm.id);
+        vmm.runtime.state = DbVmmState::SagaUnwound;
+        assert!(
+            instance_start_allowed(&logctx.log, &instance, &Some(vmm)).is_ok()
+        );
+    }
+
+    #[test]
+    fn test_instance_start_forbidden_while_creating() {
+        let logctx =
+            test_setup_log("test_instance_start_forbidden_while_creating");
+        let (mut instance, _vmm) = make_instance_and_vmm();
+        instance.runtime_state.nexus_state = DbInstanceState::Creating;
+        assert!(instance_start_allowed(&logctx.log, &instance, &None).is_err());
+    }
+
+    #[test]
+    fn test_instance_start_idempotent_if_active() {
+        let logctx = test_setup_log("test_instance_start_idempotent_if_active");
+        let (mut instance, vmm) = make_instance_and_vmm();
+        instance.runtime_state.nexus_state = DbInstanceState::Vmm;
+        instance.runtime_state.propolis_id = Some(vmm.id);
+        let mut vmm = Some(vmm);
+        vmm.as_mut().unwrap().runtime.state = DbVmmState::Starting;
+        assert!(instance_start_allowed(&logctx.log, &instance, &vmm).is_ok());
+        vmm.as_mut().unwrap().runtime.state = DbVmmState::Running;
+        assert!(instance_start_allowed(&logctx.log, &instance, &vmm).is_ok());
+        vmm.as_mut().unwrap().runtime.state = DbVmmState::Rebooting;
+        assert!(instance_start_allowed(&logctx.log, &instance, &vmm).is_ok());
+        vmm.as_mut().unwrap().runtime.state = DbVmmState::Migrating;
+        assert!(instance_start_allowed(&logctx.log, &instance, &vmm).is_ok());
```

Comment:
nitpicky: the unwraps here are a bit weird, but I guess that's a consequence of `instance_start_allowed` taking `&Option<Vmm>` rather than `Option<&Vmm>` --- if we changed it to take the latter and passed `vmm.as_ref()` when calling it elsewhere, we wouldnt need the unwraps. but, maybe not worth torturing the interface used in real code to make the tests nicer...